crates/sui-core/src/authority_aggregator.rs

```diff
@@ -1210,37 +1210,56 @@ where
                 |mut state, name, weight, result| {
                     Box::pin(async move {
                         match result {
-                            // If we are given back a certificate, then we do not need
-                            // to re-submit this transaction, we just returned the ready made
-                            // certificate. A certificate is only valid if it's formed in the
-                            // current epoch.
                             Ok(VerifiedTransactionInfoResponse {
                                 certified_transaction: Some(inner_certificate),
-                                ..
-                            }) if inner_certificate.epoch() == self.committee.epoch  => {
-                                // A validator could return a certificate from an epoch that's
-                                // different from what the authority aggregator is expecting.
-                                // In that case, we should not accept that certificate.
-                                debug!(tx_digest = ?tx_digest, name=?name.concise(), weight, "Received prev certificate from validator handle_transaction");
-                                state.certificate = Some(inner_certificate);
-                            }
-
-                            // If we didn't match the above case but here, it means that we have
-                            // a cert from a different epoch, and also have effects (i.e. already
-                            // executed), we can accept the certificate if we get 2f+1 effects.
-                            // It's an proof that the transaction has already been finalized
-                            // in a different epoch, and hence it's ok to reuse the old certificate.
-                            Ok(VerifiedTransactionInfoResponse {
                                 signed_effects: Some(inner_effects),
-                                certified_transaction: Some(inner_certificate),
                                 ..
                             }) => {
-                                if state.effects_map.add(inner_effects, weight, &self.committee) {
+                                // If we get a certificate in the same epoch, then we use it.
+                                // A certificate in a past epoch does not guaranteee finality
+                                // and validators may reject to process it.
+                                if inner_certificate.epoch() == self.committee.epoch {
+                                    debug!(tx_digest = ?tx_digest, name=?name.concise(), weight, "Received prev certificate from validator handle_transaction");
+                                    state.certificate = Some(inner_certificate);
+                                } else if inner_effects.epoch() == self.committee.epoch {
```

Comment:
This if/else is a bit strange. If a validator ever returns both certs and effects with the same epoch, you would never process the effects here.

On a side note, I realize that we stop the loop the moment we see a valid cert. This means that tracking effects is not going to be meaningful, because we never get to see 2f+1 effects anyway (we would have seen a cert way before). Due to this, I think I am ok with dropping the signed_effects case from the handle_transaction output so that we don't have to worry about it.

Reply:
When the tx is finalized from a previous epoch, the validator first [gets the tx from its perpetual store](https://github.com/MystenLabs/sui/blob/main/crates/sui-core/src/authority.rs#L2095), and then realize the effects is old so [it re-signs with the current epoch ID](https://github.com/MystenLabs/sui/blob/main/crates/sui-core/src/authority.rs#L2124). Then in the response, the certificate's epoch would be older, but effects' epoch should be current. 

In this `if else` we first check if the certificate's epoch is current, if yes, the cert is valid and we are golden. If not, then we check the signed effects.

Reply:
I'd love to have the finality proof in place sooner than later to get rid of these winding and confusing logic